### PR TITLE
Update AMDS Firmware for Timing Manager in AMDC

### LINKED ISSUE #47: Update AMDS Interface to AMDC
- [x] Create "`amds`" development branch off of `v1.3-staging` in AMDC-Firmware
- [x] Create branch off of `develop` in AMDS repo
- [x] Get working AMDC+AMDS setup and probe the interface between the boards to understand the timing
- [x] Update AMDS's firmware to send data back to AMDC's new IP driver over UART immediately after SPI communications from daughtercards are complete

Also see https://github.com/Severson-Group/AMDC-Firmware/issues/365

### PR DESCRIPTION
**Pull request by codecubepi**
This PR is a mild revision to the AMDS Firmware that makes its side of the interface compatible with the new [AMDS driver](https://github.com/Severson-Group/AMDC-Firmware/tree/v1.3-staging/ip_repo/amdc_amds_1.0) and new [Timing Manager](https://github.com/Severson-Group/AMDC-Firmware/tree/v1.3-staging/ip_repo/amdc_timing_manager_1.0) IPs introduced in v1.3 of the AMDC Firmware.

This is a relatively small code change on the AMDS firmware side. The code that is called to transmit the data from the AMDS to the AMDC driver over UART was previously in an interrupt context. Now that the data transmission is meant to happen as soon as all the daughtercards have been sampled, this code is no longer executed by its own interrupt.

Therefore, the UART transmission ISR has been moved into the function "`transmit_samples()`", which is called from the ADC sync interrupt context after sampling is completed.

Making this change also allowed me to remove the "ignore next sample" issue, since the sampling will no longer go out-of-sync  with the PWM carrier due to data transmission.

I still need to update docs.amdc.dev page on the AMDS' firmware. I will do so once we complete v1.3 in the AMDC repo.

### CODE REVIEW

**Comment by codecubepi** [comment]
Hey @codecubepi , this PR needs the following before review & close:

- [x] Test these changes with the old AMDS driver on the AMDC side for sanity sake
- [x] Test these changes with the new AMDS driver on the AMDC side for real functionality sake
- [x] Test after merging #49 into this branch
- [x] When [v1.3](https://github.com/Severson-Group/AMDC-Firmware/pull/372) in AMDC-Firmware is ready for review, this PR is also ready for final review

**Comment by npetersen2** [comment]
The default AMDS firmware in this PR works fine, but exhibits a rather large jitter in timing---I believe this issue has been present in the AMDS firmware for years. The end result is that, from the AMDC perspective, the AMDS total sampling time has around `1 us` of jitter. This was observed by @annikaolson during testing of the new `Timing Manager` feature in the AMDC FPGA.

I was doing some testing in the lab to try to figure out why and reduce the jitter. There is no reason the AMDS shouldn't behave EXACTLY the same way for each sampling instance.

I found the issue: the AMDS firmware is totally interrupt-driven: the GPIO EXTI ISR for the `SYNC_ADC` signal from the AMDC triggers the ADC sampling etc. However, there is another ISR lurking, the default `SysTick` ISR. In the current firmware, the SysTick ISR is actually higher priority vs. the SYNC_ADC ISR. Therefore, sometimes, it fires during the sampling which causes the large jitter as the MCU handles the SysTick.

Since the AMDS does not use the SysTick feature, we can simply disable the ISR on boot. Right before the `while(1)` loop in the `main.c` file, add this line of code:

```C
// Disable the SysTick ISR
//
// The SysTick ISR causes jitter in the firmware operation,
// and since this project does not use the SysTick features,
// we do not need it to run during operation!
//
// Set bit 0 to 0
SysTick->CTRL &= 0xFFFFFFFE;
``` 

Based on the scope testing, this reduces the jitter to about 70 ns (about 12 clock cycles), which is simply the MCU ISR handling jitter. Now, the AMDC won't see hardly any jitter when sampling from the AMDS. However, this fix only reduces the jitter---there is still around 10us of latency for the total AMDS sampling and data transmission.

@codecubepi can you add the line of code above to this PR and confirm the AMDS still runs per your local build? Thanks!

File: Mainboard/Firmware/motherboard_v1/Core/Src/adc.c [hunk 1]
```diff
@@ -198,19 +181,17 @@ void EXTI3_IRQHandler(void)
...
-        write_buffer_number = 1 - write_buffer_number;
-    }
+	// Switch read buffer to where we just put the new data,
+	// therefore, satisfying the property that the read_buffer_number
+	// always points to a valid set of samples!
+	read_buffer_number = 1 - read_buffer_number;
+
+	// Now that read buffer is set, we are safe to update write buffer
+	// for the next time this ISR runs...
+	write_buffer_number = 1 - write_buffer_number;
```
Comment (npetersen2) [inline Mainboard/Firmware/motherboard_v1/Core/Src/adc.c]: Tab/space issue

**Review by npetersen2 — CHANGES_REQUESTED** [review]
Thanks @codecubepi.

One comment for us to discuss---I know this code works, has been tested pretty extensively, etc. So, I hesitate to change anything. However, if you read the ADC code, it has some extra complexity with two memory buffers that it ping-pongs between due to the old ISR operation. However, now, since there is only 1 ISR context that does everything, it does not need this complexity. It would be nice to simplify the code a bit, but I am not sure we need to invest the time to do that, test it, etc.... what do you think?

**Comment by codecubepi** [comment]
> Thanks codecubepi.
> 
> One comment for us to discuss---I know this code works, has been tested pretty extensively, etc. So, I hesitate to change anything. However, if you read the ADC code, it has some extra complexity with two memory buffers that it ping-pongs between due to the old ISR operation. However, now, since there is only 1 ISR context that does everything, it does not need this complexity. It would be nice to simplify the code a bit, but I am not sure we need to invest the time to do that, test it, etc.... what do you think?

Well, if I were to fix the code to address the tab/space issue, I might as well just remove the unnecessary buffer code (and do some retesting).

Otherwise it'd be easy to open an issue and have a be a "good first issue" for future contributors... but in all reality it could turn into an issue that never really gets addressed... I think the first option of me just fixing it and retesting is probably the better way to go 🙃

**Comment by codecubepi** [comment]
@npetersen2 

I took care of removing those obselete double buffers. I ran a hardware test with the fixed [ADC data flushing](https://github.com/Severson-Group/AMDC-Firmware/issues/413), to verify that the AMDS data still matches what is seen by the AMDC's on-board ADC:

image

**Review by npetersen2 — APPROVED** [review]
So nice to have clean code :) Thanks @codecubepi

Best code is no code---115 lines removed!

**Comment by codecubepi** [comment]
> So nice to have clean code :) Thanks codecubepi
> 
> Best code is no code---115 lines removed!

Haha you're welcome. I can delete more if it'll make you even happier 😂

**Comment by codecubepi** [comment]
Note from @elsevers : We should start tagging AMDS releases with a version number that matches the AMDC Firmware version number, so user's know when firmware is compatible. So the [previous commit](https://github.com/Severson-Group/AMDS/commit/86abe4d8d466c128ab61d18d4f5d53ca8be6be7e) on `develop` would be tagged `v.1.2.x`, and this merge would get tagged as `v1.3.x`.

**Comment by npetersen2** [comment]
Per the release of AMDC-Firmware `v1.3.0`, I believe this PR should be merged, correct @codecubepi ?

**Comment by codecubepi** [comment]
> Per the release of AMDC-Firmware `v1.3.0`, I believe this PR should be merged, correct codecubepi ?

Yes, let's merge this! See my comment above about version tagging in this repo to match AMDC-Firmware.

**Comment by elsevers** [comment]
> Note from @elsevers : We should start tagging AMDS releases with a version number that matches the AMDC Firmware version number, so user's know when firmware is compatible. So the [previous commit](https://github.com/Severson-Group/AMDS/commit/86abe4d8d466c128ab61d18d4f5d53ca8be6be7e) on `develop` would be tagged `v.1.2.x`, and this merge would get tagged as `v1.3.x`.

@codecubepi, I compared notes with @npetersen2 on this and we decided on another approach to handle this. From @npetersen2 's message:

> The way you suggested could work, but I worry it will cause too much overhead, and might confuse folks to have X releases that are all exactly the same code... e.g., if the AMDS stays constant, could be easily 10+ releases for "no reason"
The way this is usually done I believe (e.g., in python) is for the AMDS to be released on its own vA.B.C numbering as makes sense, then provide a "requirements" file in the AMDC for each release saying it needs AMDS >= v2 etc.
I could see either way working, but I feel it just adds some overhead to always keep releasing the AMDS when there are no changes, plus I am not sure that codebase will change too much
>
> A future idea would be that the AMDC can query the AMDS at boot-up to determine the firmware version to ensure compatibility :slightly_smiling_face:
>
> Since, for now, this is really strictly a user/human issue, I guess keeping it in the [docs.amdc.dev](http://docs.amdc.dev/) website makes sense, probably on the "https://docs.amdc.dev/getting-started/user-guide/amds-interface/" page near the top, a new section about installing the correct AMDS firmware and a table of compatible versions between the AMDS/AMDC

**Comment by codecubepi** [comment]
@elsevers @npetersen2 

Yeah, this is a completely fair concern. I'm willing to concede 😄

**Pull request merged**